core/src/main/java/io/grpc/Context.java

```diff
@@ -434,13 +422,16 @@ public void addListener(final CancellationListener cancellationListener,
    * Remove a {@link CancellationListener}.
    */
   public void removeListener(CancellationListener cancellationListener) {
+    if (!canBeCancelled()) {
```

Comment:
Fixed canBeCancelled to be derived at construction time. It never needed a lock anyway because of immutability but it's faster now anyway.